### ISSUE TITLE
Can't run Go template

### ISSUE THREAD

**Issue by raitonoberu** [aduros/wasm4#595]
Steps to reproduce:
1. Create new Go template with `w4 new foobar --go`
2. Build it: `make`
3. Run it: `w4 run build/cart.wasm`
4. See in the console:
```
LinkError: WebAssembly.instantiate(): Import #0 module="env" function="tinygo_getCurrentStackPointer" error: function import requires a callable
```

Environment:
- Fedora 37
- wasm4 version 2.5.3
- tinygo version 0.26.0 linux/amd64 (using go version go1.19.3 and LLVM version 15.0.0)

**Comment by raitonoberu** [comment]
Downgrading tinygo to 0.25.0 fixed the issue.

**Comment by aduros** [comment]
Hi, thanks for reporting this!

I'm actually not able to repro, but I wonder if something changed with the tinygo's default [wasm.json](https://github.com/tinygo-org/tinygo/blob/release/targets/wasm.json) in 0.26, and we need to mirror any changes into the wasm4 project's target.json.

**Comment by DenialAdams** [comment]
BigAngryBob ran into other issues with the tinygo. Tinygo was invoking wasm-opt on its own, but not providing `--enable-bulk-memory`.

`[wasm-validator error in function 0] unexpected false: Bulk memory operations require bulk memory [--enable-bulk-memory], on`

His versions:
```
% w4 --version
2.5.4
% tinygo version
tinygo version 0.27.0 darwin/amd64 (using go version go1.20.4 and LLVM version 15.0.0)
% wasm-opt --version
wasm-opt version 113
```

**Comment by hrecker** [comment]
I ran into the issue mentioned above (`unexpected false: Bulk memory operations require bulk memory`) as well with the default Go template. I'm not sure how to get past to it work with Go.

**Comment by aduros** [comment]
Do you know if it's failing on `tinygo` compilation, or the `wasm-opt` post-build pass?

**Comment by hchac** [comment]
As @DenialAdams mentioned, tinygo is internally calling wasm-opt with its own flags. This error happens during the tinygo compilation phase, before the `wasm-opt` post-build pass.

I was able to compile successfully by replacing the template's `target.json` with the one from the tinygo project: https://github.com/tinygo-org/tinygo/blob/release/targets/wasm.json (this one contains the `-mbulk-memory` flag, which bypasses the current compiling error)

However, I am now facing the issue: `Uncaught (in promise) TypeError: import object field 'wasi_snapshot_preview1' is not an Object` when I open the game in the browser through `w4 run build/cart.wasm`. Not sure how to fix that one.

I'm using go `1.21.1`, tinygo `0.30.0`, and w4 `2.5.4`.

tl;dr

1) Replace contents of `target.json` in template with https://github.com/tinygo-org/tinygo/blob/release/targets/wasm.json
2) Remove `wasm-opt` steps in Makefile, because tinygo already calls `wasm-opt` internally
3) Since tinygo runs `wasm-opt` internally, the user needs to have `wasm-opt` installed
    3.1) If tinygo was installed through a package manager, then a separate install of `binaryen` is needed to get `wasm-opt`
    3.2) If tinygo was installed by downloading it from their GitHub releases, then `wasm-opt` already comes included in their `bin` folder. tinygo will automatically find this `wasm-opt` in their own code: https://github.com/tinygo-org/tinygo/blob/731532cd2b6353b60b443343b51296ec0fafae09/goenv/goenv.go#L143-L165

**Comment by hchac** [comment]
Replacing all of `target.json` was probably too intrusive on my part. The contentious flag in the current `target.json` seems to be `--strip-all` https://github.com/aduros/wasm4/blob/aca37f98e509f96faa789efb9df8925d28f65bbf/cli/assets/templates/go/target.json#L21

Looks like this flag isn't playing nicely with the latest `tinygo`. If you remove that flag you can get past the `unexpected false: Bulk memory operation (bulk memory is disabled)` build errors. However, you still have the `Uncaught (in promise) TypeError: import object field 'wasi_snapshot_preview1' is not an Object` error in the console once you run the cart.

Unfortunately I don't currently know anything about tinygo or wasm to figure out if any of these changes are lining up to produce a correct wasm4 cart.

**Comment by aduros** [comment]
I haven't had time to look at this, but more experienced Go developers may have some insight. Ping @peterhellberg @christopher-kleine

**Comment by christopher-kleine** [comment]
There's a small game I'd like to implement using Go anyway. I'll look into this in the next days, I guess. I'm currently busy with work.

**Comment by peterhellberg** [comment]
@christopher-kleine Cool, I have primarily used [Zig](https://ziglang.org/) :zap: lately for fantasy console related stuff, so have no up to date experience with [TinyGo](https://tinygo.org/) together with [WASM4](https://wasm4.org/).

**Comment by christopher-kleine** [comment]
So far I was able to confirm the problem. I tried the default target.json supplied by WASM4 and also the target.json directly from TinyGo. Both failed.

Digging a little, it's stated, that we need the wasm_exec.js from TinyGo instead of the one from Go itself.
I *tried* this, but was unsuccessful as of now. I might have to ask on their Repo about this.

**Comment by PrimalPimmy** [comment]
This is still an issue, was trying the same thing and same error on the console. Any leads?

**Comment by christopher-kleine** [comment]
Sorry, I didn't find anything new. Currently I don't have the time to investigate further, since other things popped up.

My best advice so far would be: Either try to downgrade to an earlier version of Tinygo or - if an older version of WASM4 is fine - try the [Docker Images](https://hub.docker.com/r/anihex/wasm4).

Reminds me: Should I take them down, @aduros ?
I can't maintain the docker images for quite some time, I'm afraid. And I don't want people come running to you because of bugs in those older versions.

**Comment by 0riginaln0** [comment]
ok so for now here is how to use wasm-4 with go without docker image

1. installing proper version of tinygo (0.21.0)
https://github.com/tinygo-org/tinygo/releases/tag/v0.21.0

2. installing proper go version (1.17)
I had go 1.22.1 installed globally. I installed the second version of go:
run:
```go install golang.org/dl/go1.17@latest```
find your GOPATH with `go env`
explicitly add GOPATH/bin path to the `~/.bashrc`
It is needed because go install installs go versions in the gopath/bin folder
run:
```go1.17 download```

3. install wasm-opt
https://github.com/WebAssembly/binaryen/releases
just unpack the archive and add /bin to the path

4. run `go1.17 env` and add to the path its GOROOT and GOPATH

At the end you will have your `~/.bashrc` something similar to this:
```export PATH=$PATH:/home/ryzh/Downloads/Programs/w4/w4-linux
export PATH=$PATH:/home/ryzh/go/bin
export PATH=$PATH:/home/ryzh/Downloads/Programs/w4/binaryen-version_117/bin
export GOROOT=/home/ryzh/sdk/go1.17
export GOPATH=/home/ryzh/go```

Now you can run w4 commands and all will (hopefully) work.

When you finish your w4 things you can comment out GOROOT and GOPATH. So after that you can continue use your main go version
```#export GOROOT=/home/ryzh/sdk/go1.17
#export GOPATH=/home/ryzh/go```

**Comment by AlRado** [comment]
It seems that the problem with the latest versions of Go and TinyGo is still in the incorrect target.json. With these settings, Wasm-4 works for me without errors. In my case, versions **Go 1.22.2** and **TinyGo 0.31.2**

**target.json** content:
_{
    "llvm-target": "wasm32-unknown-unknown",
    "cpu": "generic",
    "features": "+mutable-globals,+nontrapping-fptoint,+sign-ext,+bulk-memory",
    "build-tags": [
        "tinygo.wasm",
        "wasm_unknown"
    ],
    "goos": "linux",
    "goarch": "arm",
    "linker": "wasm-ld",
    "rtlib": "compiler-rt",
    "scheduler": "none",
    "cflags": [
        "-mno-bulk-memory",
        "-mnontrapping-fptoint",
        "-msign-ext"
    ],
    "ldflags": [
        "--allow-undefined",
        "--no-demangle",
        "--import-memory",
        "--initial-memory=65536",
        "--max-memory=65536",
        "--stack-first",
        "--no-entry",
        "-zstack-size=14752"
    ]
}_

**Comment by christopher-kleine** [comment]
I can partially confirm this. The "Hello World" example works.
But I also tried to compile the snake demo and my Tic-Tac-Toe source.
Both *compile* just fine. But I can't run them.

```
Uncaught (in promise) LinkError: import object field 'tinygo_getCurrentStackPointer' is not a Function
```

**Comment by 0riginaln0** [comment]
For the windows users, the easiest way is to write in powershell:
1. `scoop install go@1.17`
2. `scoop install tinygo@0.21.0`
4. `scoop install binaryen`
5. `go env` -> and add both GOROOT and GOPATH into the enviroment variables

**Comment by rencrawford** [comment]
I got the snake and hello world example to work with the following config: 

tinygo version 0.33.0 
windows/amd64 
go version go1.23.1 
LLVM version 18.1.2

Build Size  12557kb
![Screenshot 2024-09-19 234340](https://github.com/user-attachments/assets/7832c06a-7272-4790-9fa4-fa7a75a11e9e)

# target.json fix 
{
  "llvm-target": "wasm32--wasi",
  "build-tags": [ "tinygo.wasm" ],
  "goos": "js",
  "goarch": "wasm",
  "linker": "wasm-ld",
  "libc": "wasi-libc",
  "cflags": [
    "--target=wasm32--wasi",
    "--sysroot={root}/lib/wasi-libc/sysroot",
    "-Oz"
  ],

  "ldflags": [
    "--allow-undefined",
    "--no-demangle",
    "--import-memory",
    "--initial-memory=65536",
    "--max-memory=65536",
    "--stack-first",
    "-zstack-size=14752"
  ],
  "wasm-abi": "js"
}

# build command 
tinygo build -target ./target.json -panic trap -opt=z -scheduler=none -gc=leaking -no-debug -o build/cart.wasm .
w4 run build/cart.wasm

**Comment by 0riginaln0** [comment]
> # build command
> tinygo build -target ./target.json -panic trap -opt=z -scheduler=none -gc=leaking -no-debug -o build/cart.wasm . w4 run build/cart.wasm

-gc=leaking seems dangerous
